From f9b6af4f64e58205027e043fe624d076d43fc1ba Mon Sep 17 00:00:00 2001
From: Yang Chang <yang.chang@intel.com>
Date: Mon, 22 Apr 2013 11:37:49 +0800
Subject: Dalvik: Make sure Xgc:noprecise always work

BZ: 101896

visitThreadStack should be stricter while deciding if a method frame
should be scanned with precise GC info. Currently, if there's a
RegisterMap, it will be used even if gDvm.preciseGc (-Xgc:noprecise)
is false.

Category: device-enablement
Domain: AOSP-Dalvik-GC
Origin: internal
Upstream-Candidate: yes

Change-Id: Ie1b3198295360e3903126054fc356d92d2f1e40a
Orig-MCG-Change-Id: I4e452e3015a21ef73941b144cedd52d8a9fbf214
Signed-off-by: Yang Chang <yang.chang@intel.com>
Signed-off-by: Qiming Shi <qiming.shi@intel.com>
Signed-off-by: Serguei Katkov <serguei.i.katkov@intel.com>
---
 vm/alloc/Copying.cpp |    2 +-
 vm/alloc/Visit.cpp   |    2 +-
 2 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/vm/alloc/Copying.cpp b/vm/alloc/Copying.cpp
index 74cb6b8..c50838d 100644
--- a/vm/alloc/Copying.cpp
+++ b/vm/alloc/Copying.cpp
@@ -1605,7 +1605,7 @@ static void scavengeThreadStack(Thread *thread)
 
             //LOG_SCAV("PGC: %s.%s", method->clazz->descriptor, method->name);
 
-            if (pMap != NULL) {
+            if (gDvm.preciseGc && pMap != NULL) {
                 /* found map, get registers for this address */
                 int addr = saveArea->xtra.currentPc - method->insns;
                 regVector = dvmRegisterMapGetLine(pMap, addr);
diff --git a/vm/alloc/Visit.cpp b/vm/alloc/Visit.cpp
index a04a835..9339838 100644
--- a/vm/alloc/Visit.cpp
+++ b/vm/alloc/Visit.cpp
@@ -95,7 +95,7 @@ static void visitThreadStack(RootVisitor *visitor, Thread *thread, void *arg)
         if (method != NULL && !dvmIsNativeMethod(method)) {
             const RegisterMap* pMap = dvmGetExpandedRegisterMap(method);
             const u1* regVector = NULL;
-            if (pMap != NULL) {
+            if (gDvm.preciseGc && pMap != NULL) {
                 /* found map, get registers for this address */
                 int addr = saveArea->xtra.currentPc - method->insns;
                 regVector = dvmRegisterMapGetLine(pMap, addr);
-- 
1.7.4.1

